designs/sam-debugging/local-sam-debugging.md

```diff
@@ -0,0 +1,308 @@
+-   Region
+
+## <a id="debugging-experiences"></a> Local Debugging Experiences
+
+### Debug Configurations
+
+Debug Configurations are the idiomatic approach to running and debugging software in VS Code. They are also a reusable component - the Toolkit is able to internally produce and execute these configurations on the fly (for example as a part of the [CodeLenses](#codelenses) functionality). This is the Toolkit's main experience for debugging SAM Template resources.
+
+The Toolkit implements a Debug Configuration type `aws-sam`. Users can author and maintain these configuration entries, then launch them by pressing F5. When launched, this configuration type:
+
+-   validates debug configuration inputs
+-   uses SAM CLI to build a SAM Application
+-   uses SAM CLI to invoke a SAM Template Resource
+-   attaches a debugger to the SAM invocation
+-   if the debug configuration is for a local api gateway invoke, the debugger is detached after the http request is made, but SAM CLI remains active. The debug configuration implementation terminates the SAM CLI session to prevent a proliferation of CLI processes.
+
+In the most basic form, the debug configuration references a SAM Template file location, and a Resource within that file. Other execution parameters can be configured, but are optional.
+
+Debugging local lambda invokes and local api gateway invokes each require slightly different inputs. The `aws-sam` Debug Configuration uses different request types to accommodate these variations.
+
+These debug configurations are authored in a json file. The following Toolkit assistance is provided:
+
+-   autocompletion with descriptions is provided for `aws-sam` related fields
+    -   There is no autocompletion available for specific values in a configuration. For example, if a user types in the location of a SAM Template file, there is no filesystem-based autocompletion. The Debug Configuration validates the configuration and notifies of errant values when it is run.
+-   snippets to produce typical (or starter) `aws-sam` debug configurations
+-   when no launch.json file is present in a workspace, VS Code exposes functionality that allows users to request auto-generated Debug Configurations. In this situation, the toolkit generates an `aws-sam` Debug Configuration for all `AWS::Serverless::Function` Resources detected within all SAM Templates located in the workspace.
+
+Example Debug Configuration entries can be found in the [Appendix](#sample-debug-configurations)
+
+Standalone Lambda function handlers are not supported through Debug Configurations.
+
+### <a id="codelenses"></a> CodeLenses
+
+#### CodeLenses in SAM Template files
+
+The following CodeLenses appear above every template resource of type `AWS::Serverless::Function`:
+
+-   Run Locally - See below for details
+-   Debug Locally - See below for details
+-   Configure - allows the user to configure a limited set of arguments that are used with the Run and Debug CodeLenses
```

Comment:
I don't think that Run|Debug|Configure CodeLenses on the SAM Templates add functional value over having Debug Configurations. There's no need to have a separately configured Debug flow here. Instead, what I'd like to do is display codelenses for any debug configurations that reference this template + resource, and include the "Add Debug Config" CodeLens from the next bullet point.

This should increase consistency, and align with some of @bryceitoc9 remarks